src/compiler.ts

```diff
@@ -7975,24 +8148,138 @@ export class Compiler extends DiagnosticEmitter {
+        contextualTypeArguments
+      );
+      this.compileFunction(instance)
+    }
+    this.currentType = instance.signature.type;
+
     var index = this.ensureFunctionTableEntry(instance); // reports
-    return index < 0
-      ? this.module.unreachable()
-      : this.module.i32(index);
+
+    if(index < 0) return this.module.unreachable();
+
+    if(instance.closedLocals.size > 0) {
+      //Create field declarations for the function and each closed local
+      var members = Array<DeclarationStatement>(instance.closedLocals.size + 1);
+      members[0] = this.program.makeNativeMember("__functionPtr", "u32")
+      for (let _values = Map_values(instance.closedLocals), i = 0, k = _values.length; i < k; ++i) {
+        let local = unchecked(_values[i]);
+        members[i + 1] = this.program.makeNativeMember(local.name, local.type.intType.toString())
+      }
+
+      //Create a native class prototype with a dummy syntax tree, similar to native functions
+      var closureClassPrototype = assert(this.program.makeNativeClassPrototype(
+        "closure|" + this.program.nextClassId.toString(),
+        members
+      ));
+
+      //Resolve this prototype to get the class
+      var closureClass = this.resolver.resolveClass(closureClassPrototype, null)!;
+
+      //Compile this class to get the type
+      this.compileClass(closureClass);
+
+      //Append the appropriate signature and flags for this closure type, then set it to currentType
+      this.currentType = closureClass.type.asClosure(instance.signature)
+
+      //create a local which will hold our closure class instance
+      var tempLocal = flow.getAutoreleaseLocal(this.currentType);
+
+      //copied closed locals into type
+      this.currentType.locals = instance.closedLocals;
+
+      var closureExpr = this.module.flatten([
+        this.module.local_set( //Allocate memory for the closure
+          tempLocal.index,
+          this.makeRetain(
+            this.module.call(this.program.allocInstance.internalName, [
+              this.module.i32(closureClass.type.byteSize),
+              this.module.i32(closureClass.id)
+            ], NativeType.I32)
+          )
+        ),
+        this.module.store( //Store the function pointer at the first index
+          4,
+          this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+          this.module.i32(index),
+          NativeType.I32,
+          0
+        ),
+        this.module.local_get(tempLocal.index, this.options.nativeSizeType) //load the closure locals index
+      ], this.options.nativeSizeType);
+
+      //flow.freeTempLocal(tempLocal);
+
+      return closureExpr;
+    }
+
+    return this.module.i32(index);
+  }
+
+  private ifClosure(
+    indexExpr: ExpressionRef,
+    thenExpr: ExpressionRef,
+    elseExpr: ExpressionRef
+  ): ExpressionRef {
+    var module = this.module;
+
+    return module.if(
+      module.binary(
+        BinaryOp.EqI32,
```

Comment:
`indexExpr` is of type `usize` here, so requires separate instructions for 32-bit and, eventually, 64-bit Wasm. Ofc this isn't a thing as of today, but other code already takes this into account so we don't have to rewrite half the compiler and go bug hunting for an extended period of time once 64-bit Wasm hits. Typically checked via `this.options.isWasm64`, sometimes also by comparing `nativeSizeType == NativeType.I64` where code uses that already.